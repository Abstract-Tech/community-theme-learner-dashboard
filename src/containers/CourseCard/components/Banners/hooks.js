import { useIntl } from '@edx/frontend-platform/i18n';
import { useCardValues } from 'hooks';
import { selectors } from 'data/redux';

const { cardData } = selectors;

export const useCourseBannerData = ({ courseNumber }) => ({
  courseData: useCardValues(courseNumber, {
    isVerified: cardData.isVerified,
    isCourseRunActive: cardData.isCourseRunActive,
    canUpgrade: cardData.canUpgrade,
    isAuditAccessExpired: cardData.isAuditAccessExpired,
    courseWebsite: cardData.courseWebsite,
  }),
  formatMessage: useIntl().formatMessage,
});

export default {
  useCourseBannerData,
};
