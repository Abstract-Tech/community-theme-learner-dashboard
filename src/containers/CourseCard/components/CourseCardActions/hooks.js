import { useDispatch } from 'react-redux';

import { Locked } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks as appHooks } from 'data/redux';
import messages from './messages';

export const useCardActionData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const dispatch = useDispatch();
  const {
    canUpgrade,
    coursewareAccess,
    hasStarted,
    isAudit,
    isAuditAccessExpired,
    isVerified,
  } = appHooks.useCardEnrollmentData(cardId);
  const { isArchived } = appHooks.useCardCourseRunData(cardId);
  const {
    canChange,
    hasSessions,
    isEntitlement,
    isExpired,
    isFulfilled,
  } = appHooks.useCardEntitlementData(cardId);
  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  const hasAccess = coursewareAccess.isStaff || !(
    coursewareAccess.hasUnmetPrereqs
    || coursewareAccess.isTooEarly
    || isArchived);

  const selectSessionButton = {
    children: formatMessage(messages.selectSession),
    onClick: openSessionModal,
    disabled: !hasAccess || (!canChange || !hasSessions),
  };

  const viewCourseButton = {
    children: formatMessage(messages.viewCourse),
    disabled: !hasAccess || (isEntitlement && isExpired),
  };

  const beginCourseButton = {
    children: formatMessage(messages.beginCourse),
    disabled: !hasAccess,
  };

  const resumeButton = {
    children: formatMessage(messages.resume),
    disabled: isAudit && isAuditAccessExpired,
  };

  const upgradeButton = {
    iconBefore: Locked,
    variant: 'outline-primary',
    children: formatMessage(messages.upgrade),
    disabled: !canUpgrade,
  };

  let primary;
  if (isEntitlement) {
    primary = isFulfilled ? selectSessionButton : viewCourseButton;
  } else if (!hasStarted) {
    primary = beginCourseButton;
  } else if (isArchived) {
    primary = viewCourseButton;
  } else {
    primary = resumeButton;
  }

  const secondary = (isEntitlement || !isVerified) ? null : upgradeButton;
  return { primary, secondary };
};

export default useCardActionData;
