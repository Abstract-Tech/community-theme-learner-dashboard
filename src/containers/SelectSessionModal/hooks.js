import { useDispatch } from 'react-redux';

import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks as appHooks } from 'data/redux';

import messages from './messages';

export const useSelectSessionModalData = () => {
  const dispatch = useDispatch();
  const selectedCardId = appHooks.useSelectSessionModalData().cardId;

  const {
    entitlementSessions,
    isFulfilled,
  } = appHooks.useCardEntitlementsData(selectedCardId);

  const { title: courseTitle } = appHooks.useCardCourseData(selectedCardId);

  const { formatMessage } = useIntl();

  let header;
  let hint;
  if (isFulfilled) {
    header = formatMessage(messages.changeOrLeaveHeader);
    hint = formatMessage(messages.changeOrLeaveHint);
  } else {
    header = formatMessage(messages.selectSessionHeader, {
      courseTitle,
    });
    hint = formatMessage(messages.selectSessionHint);
  }
  const updateCallback = appHooks.useUpdateSelectSessionModalCallback;

  return {
    showModal: selectedCardId != null,
    closeSessionModal: updateCallback(dispatch, null),
    openSessionModal: (cardId) => updateCallback(dispatch, cardId),
    showLeaveOption: isFulfilled,
    entitlementSessions,
    hint,
    header,
  };
};

export default useSelectSessionModalData;
