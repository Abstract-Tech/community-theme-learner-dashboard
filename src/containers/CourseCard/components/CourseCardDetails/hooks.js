import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as module from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage, formatDate } = useIntl();
  const enrollment = appHooks.useCardEnrollmentData(cardId);
  const courseRun = appHooks.useCardCourseRunData(cardId);
  if (!courseRun.isStarted) {
    return formatMessage(messages.courseStarts, { startDate: courseRun.startDate });
  }
  if (enrollment.isEnrolled) {
    if (enrollment.isAudit) {
      const {
        accessExpirationDate,
        isAuditAccessExpired,
      } = enrollment;
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    const { isArchived, endDate } = courseRun;
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ dispatch, cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(cardId).name;
  const { courseNumber } = appHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementsData(cardId);

  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: module.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    formatMessage,
    courseNumber,
  };
};

export default useCardDetailsData;
