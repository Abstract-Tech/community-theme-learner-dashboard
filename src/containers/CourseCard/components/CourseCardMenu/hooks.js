import React from 'react';
import { StrictDict } from 'utils';
import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val),
  isEmailSettingsVisible: (val) => React.useState(val),
});

export const unenrollModalHooks = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const emailSettingsModalHooks = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const menuHooks = () => {
  const unenrollModal = module.unenrollModalHooks();
  const emailSettingsModal = module.emailSettingsModalHooks();
  return {
    emailSettingsModal,
    unenrollModal,
  };
};

export default menuHooks;
