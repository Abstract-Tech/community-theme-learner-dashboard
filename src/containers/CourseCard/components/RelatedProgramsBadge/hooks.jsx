import React from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';

import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import messages from './messages';
import * as module from './hooks';

export const state = StrictDict({
  isOpen: (val) => React.useState(val), // eslint-disable-line
});

export const useRelatedProgramsBadgeData = ({ cardId }) => {
  const [isOpen, setIsOpen] = module.state.isOpen(false);
  const { formatMessage } = useIntl();
  const numPrograms = appHooks.useCardRelatedProgramsData(cardId).length;
  let programsMessage = '';
  if (numPrograms) {
    programsMessage = formatMessage(
      numPrograms === 1 ? messages.badgeLabelSingular : messages.badgeLabelPlural,
      { numPrograms },
    );
  }

  return {
    numPrograms,
    programsMessage,
    isOpen,
    openModal: React.useCallback(() => setIsOpen(true), [setIsOpen]),
    closeModal: React.useCallback(() => setIsOpen(false), [setIsOpen]),
  };
};

export default useRelatedProgramsBadgeData;
