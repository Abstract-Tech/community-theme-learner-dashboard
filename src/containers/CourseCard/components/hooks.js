import { reduxHooks } from 'hooks';
import { EXECUTIVE_EDUCATION_COURSE_MODES } from '../../../data/constants/course';

export const useActionDisabledState = (cardId) => {
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const {
    canUpgrade, hasAccess, isAudit, isAuditAccessExpired, mode,
  } = reduxHooks.useCardEnrollmentData(cardId);
  const {
    isEntitlement, isFulfilled, canChange, hasSessions,
  } = reduxHooks.useCardEntitlementData(cardId);

  const { resumeUrl, homeUrl, upgradeUrl } = reduxHooks.useCardCourseRunData(cardId);

  const disableBeginCourse = !homeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableResumeCourse = !resumeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableViewCourse = !hasAccess || (isAudit && isAuditAccessExpired);
  const disableSelectSession = !isEntitlement || isMasquerading || !hasAccess || (!canChange || !hasSessions);
  const disableUpgradeCourse = !upgradeUrl || (isMasquerading && !canUpgrade);

  const disableCourseTitle = (isEntitlement && !isFulfilled) || disableViewCourse;

  const isExecutiveEd2uCourse = (EXECUTIVE_EDUCATION_COURSE_MODES.includes(mode));

  return {
    disableBeginCourse,
    disableResumeCourse,
    disableViewCourse,
    disableUpgradeCourse,
    disableSelectSession,
    disableCourseTitle,
    isExecutiveEd2uCourse,
  };
};

export default useActionDisabledState;
