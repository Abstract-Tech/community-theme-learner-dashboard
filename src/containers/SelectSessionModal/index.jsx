import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';
import {
  ActionRow,
  Button,
  Form,
  ModalDialog,
} from '@edx/paragon';

import { nullMethod } from 'hooks';
import { dateFormatter } from 'utils';

import useSelectSessionModalData from './hooks';
import messages from './messages';

export const SelectSessionModal = () => {
  const {
    entitlementSessions,
    showModal,
    closeSessionModal,
    showLeaveOption,
    header,
    hint,
  } = useSelectSessionModalData();

  const { formatMessage, formatDate } = useIntl();

  return (
    <ModalDialog
      isOpen={showModal}
      onClose={nullMethod}
      hasCloseButton={false}
      size="md"
      className="p-4 px-4.5"
      title={header}
    >
      <h3>{header}</h3>
      <Form.Group className="pt-3">
        <Form.Label>{hint}</Form.Label>
        <Form.RadioSet name="sessions" className="pt-3 pb-4">
          {entitlementSessions?.map((session) => (
            <Form.Radio key={session.startDate} value={session.startDate}>
              {dateFormatter(formatDate, session.startDate)} - {dateFormatter(formatDate, session.endDate)}
            </Form.Radio>
          ))}
          {showLeaveOption && (
            <Form.Radio value="leave">
              {formatMessage(messages.leaveSessionOption)}
            </Form.Radio>
          )}
        </Form.RadioSet>
      </Form.Group>
      <ActionRow>
        <Button variant="tertiary" onClick={closeSessionModal}>
          {formatMessage(messages.nevermind)}
        </Button>
        <Button>{formatMessage(messages.confirmSession)}</Button>
      </ActionRow>
    </ModalDialog>
  );
};

export default SelectSessionModal;
