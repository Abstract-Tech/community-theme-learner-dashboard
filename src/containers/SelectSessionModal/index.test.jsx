import React from 'react';
import { shallow } from 'enzyme';

import hooks from './hooks';
import SelectSessionModal from '.';

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const hookReturn = {
  entitlementSessions: [],
  showModal: true,
  closeSessionModal: jest.fn().mockName('useSelectSessionModalData.closeSessionModal'),
  showLeaveOption: true,
  header: 'test-header',
  hint: 'test-hint',
};

const courseNumber = 'my-test-course-number';

const availableSessions = [
  { startDate: '1/2/2000', endDate: '1/2/2020', courseNumber },
  { startDate: '2/3/2000', endDate: '2/3/2020', courseNumber },
  { startDate: '3/4/2000', endDate: '3/4/2020', courseNumber },
];

describe('SelectSessionModal', () => {
  describe('snapshot', () => {
    test('empty modal with leave option ', () => {
      hooks.mockReturnValueOnce({
        ...hookReturn,
      });
      expect(shallow(<SelectSessionModal />)).toMatchSnapshot();
    });

    test('modal with leave option ', () => {
      hooks.mockReturnValueOnce({
        ...hookReturn,
        entitlementSessions: [...availableSessions],
      });
      expect(shallow(<SelectSessionModal />)).toMatchSnapshot();
    });

    test('modal without leave option ', () => {
      hooks.mockReturnValueOnce({
        ...hookReturn,
        entitlementSessions: [...availableSessions],
        showLeaveOption: false,
      });
      expect(shallow(<SelectSessionModal />)).toMatchSnapshot();
    });
  });
});
