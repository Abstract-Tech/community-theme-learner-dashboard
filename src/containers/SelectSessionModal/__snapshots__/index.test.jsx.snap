// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectSessionModal snapshot empty modal with leave option  1`] = `
<ModalDialog
  className="p-4 px-4.5"
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  size="md"
>
  <h3>
    test-header
  </h3>
  <Form.Group
    className="pt-3"
  >
    <Form.Label>
      test-hint
    </Form.Label>
    <Form.RadioSet
      className="pt-3 pb-4"
      name="sessions"
    >
      <Form.Radio
        value="leave"
      >
        Leave session
      </Form.Radio>
    </Form.RadioSet>
  </Form.Group>
  <ActionRow>
    <Button
      onClick={[MockFunction useSelectSessionModalData.closeSessionModal]}
      variant="tertiary"
    >
      Nevermind
    </Button>
    <Button>
      Confirm Session
    </Button>
  </ActionRow>
</ModalDialog>
`;

exports[`SelectSessionModal snapshot modal with leave option  1`] = `
<ModalDialog
  className="p-4 px-4.5"
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  size="md"
>
  <h3>
    test-header
  </h3>
  <Form.Group
    className="pt-3"
  >
    <Form.Label>
      test-hint
    </Form.Label>
    <Form.RadioSet
      className="pt-3 pb-4"
      name="sessions"
    >
      <Form.Radio
        key="1/2/2000"
        value="1/2/2000"
      >
        1/2/2000
         - 
        1/2/2020
      </Form.Radio>
      <Form.Radio
        key="2/3/2000"
        value="2/3/2000"
      >
        2/3/2000
         - 
        2/3/2020
      </Form.Radio>
      <Form.Radio
        key="3/4/2000"
        value="3/4/2000"
      >
        3/4/2000
         - 
        3/4/2020
      </Form.Radio>
      <Form.Radio
        value="leave"
      >
        Leave session
      </Form.Radio>
    </Form.RadioSet>
  </Form.Group>
  <ActionRow>
    <Button
      onClick={[MockFunction useSelectSessionModalData.closeSessionModal]}
      variant="tertiary"
    >
      Nevermind
    </Button>
    <Button>
      Confirm Session
    </Button>
  </ActionRow>
</ModalDialog>
`;

exports[`SelectSessionModal snapshot modal without leave option  1`] = `
<ModalDialog
  className="p-4 px-4.5"
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  size="md"
>
  <h3>
    test-header
  </h3>
  <Form.Group
    className="pt-3"
  >
    <Form.Label>
      test-hint
    </Form.Label>
    <Form.RadioSet
      className="pt-3 pb-4"
      name="sessions"
    >
      <Form.Radio
        key="1/2/2000"
        value="1/2/2000"
      >
        1/2/2000
         - 
        1/2/2020
      </Form.Radio>
      <Form.Radio
        key="2/3/2000"
        value="2/3/2000"
      >
        2/3/2000
         - 
        2/3/2020
      </Form.Radio>
      <Form.Radio
        key="3/4/2000"
        value="3/4/2000"
      >
        3/4/2000
         - 
        3/4/2020
      </Form.Radio>
    </Form.RadioSet>
  </Form.Group>
  <ActionRow>
    <Button
      onClick={[MockFunction useSelectSessionModalData.closeSessionModal]}
      variant="tertiary"
    >
      Nevermind
    </Button>
    <Button>
      Confirm Session
    </Button>
  </ActionRow>
</ModalDialog>
`;
